modules/renter/hostdb/hostentry.go

```diff
@@ -113,19 +115,43 @@ func (hdb *HostDB) AverageContractPrice() types.Currency {
 	return totalPrice.Div64(uint64(len(hosts)))
 }
 
-// IsOffline reports whether a host is offline. If the HostDB has no record of
-// the host, IsOffline will return false.
-//
-// TODO: Is this behavior that makes sense?
+// IsOffline reports whether h is offline. A host is considered offline if it
+// has been scanned at least once in the last three days, but hasn't responded
+// to the scan during that period. If the host has not been scanned in the last
+// three days, IsOffline will scan it, so the caller should treat this call as
+// blocking. If the host is not present in the HostDB, IsOffline returns false.
 func (hdb *HostDB) IsOffline(addr modules.NetAddress) bool {
+	// lookup entry
 	hdb.mu.RLock()
-	defer hdb.mu.RUnlock()
-
-	if _, ok := hdb.activeHosts[addr]; ok {
+	var lastSeen, lastScanned time.Time
+	entry, ok := hdb.allHosts[addr]
+	if ok {
+		lastSeen, lastScanned = entry.LastSeen, entry.LastScanned
+	}
+	hdb.mu.RUnlock()
+	if !ok {
 		return false
 	}
-	if h, ok := hdb.allHosts[addr]; ok {
-		return !h.Online
+
+	if time.Since(lastScanned) > uptimeThreshold {
```

Comment:
By implement it this way you are requiring the contractor to call it in a background thread, and I don't think our implementation of the contractor is stable enough to justify limitations in upstream.

IsOffline is a query function, I don't think I would expect a function like that to block for network calls. I would generally expect it to just look at the existing state.